# Trim adapter sequence and require minimum read length
### cutadapt version 1.15
```bash
cutadapt --adapter AGATCGGAAGAGCACACGTCTGAACTCCAGTCA -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT --output=sample.R1.trimmed.fastq --paired-output=sample.R2.trimmed.fastq --pair-filter=any --minimum-length=20 sample.R1.fastq sample.R2.fastq
```

# Build genome assembly index
### STAR 2.7.1a
```bash
STAR --runMode genomeGenerate --runThreadN 8 --genomeDir cascadePacBioMasked --genomeFastaFiles genome.fasta
```

# Align RNAseq with STAR
### STAR 2.7.1a
```bash
STAR --runMode alignReads --genomeLoad NoSharedMemory --soloStrand Forward --outSAMtype BAM SortedByCoordinate --genomeDir /path/to/cascadePacBioMasked --readFilesIn sample.R1.trimmed.fastq sample.R2.trimmed.fastq --outFileNamePrefix cascadeAlignments
```

# Create finalOutFileList.txt 
### *final.out files are generated by STAR
``` bash
ls -1 *final.out > finalOutFileList.txt
```

# Get mapping rate of reads to the reference assembly
``` bash
python parseStarFinalOutFile.py finalOutFileList.txt > mappingRate.txt
```

# Assemble transcripts with StringTie
### StringTie v1.3.3b
``` bash
stringtie sample.sortedByCoord.out.bam -j 2 -e -G hopCascadeAugustusGeneModels.gff -o sample.sortedByCoord.out.gtf --fr -A sample.sortedByCoord.out.tab
```

# Create samplelist.txt
``` bash
ls -1 *gtf > samplelist.txt
```

# Create sampleIDs.txt file for prepDE.py
``` bash
python changeID.py samplelist.txt 
```

# Create files for differential gene expression analysis with DESeq2
### prepDE.py script obtained from StringTie: https://github.com/gpertea/stringtie/blob/master/prepDE.py
``` bash
python prepDE.py -i sampleIDs.txt
```

# Create design.csv file
``` bash
python createDesignCSV.py gene_count_matrix.csv 
```
