### Trim adapter sequence and require minimum read length
### cutadapt version 1.15
```bash
cutadapt --adapter AGATCGGAAGAGCACACGTCTGAACTCCAGTCA -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT --output=sample.R1.trimmed.fastq --paired-output=sample.R2.trimmed.fastq --pair-filter=any --minimum-length=20 sample.R1.fastq sample.R2.fastq
```

# build genome assembly index
STAR --runMode genomeGenerate --runThreadN 8 --genomeDir cascadePacBioMasked --genomeFastaFiles genome.fasta

# align RNAseq with STAR
STAR --runMode alignReads --genomeLoad NoSharedMemory --soloStrand Forward --outSAMtype BAM SortedByCoordinate --genomeDir /path/to/cascadePacBioMasked --readFilesIn sample.R1.trimmed.fastq sample.R2.trimmed.fastq --outFileNamePrefix cascadeAlignments

# get finalOutFileList.txt --> *final.out files generated by STAR
ls -1 *final.out > finalOutFileList.txt

# get mapping rate of reads to the reference assembly
python parseStarFinalOutFile.py finalOutFileList.txt > mappingRate.txt

# assemble transcripts with stringtie
stringtie sample.sortedByCoord.out.bam -j 2 -e -G hopCascadeAugustusGeneModels.gff -o sample.sortedByCoord.out.gtf --fr -A sample.sortedByCoord.out.tab

# get samplelist.txt
ls -1 *gtf > samplelist.txt

# create sampleIDs.txt file for prepDE.py
python changeID.py samplelist.txt 

# create files for differential gene expression analysis
python prepDE.py -i sampleIDs.txt

# create design.csv file
python createDesignCSV.py gene_count_matrix.csv 
